bootloader/piboot.go

```diff
@@ -0,0 +1,395 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootloader
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader/ubootenv"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// sanity - piboot implements the required interfaces
+var (
+	_ Bootloader                             = (*piboot)(nil)
+	_ ExtractedRecoveryKernelImageBootloader = (*piboot)(nil)
+)
+
+const (
+	pibootCfgFilename = "piboot.conf"
+	pibootPartFolder  = "/piboot/ubuntu/"
+	// TODO Use boot.InitramfsUbuntuSeedDir
+	// Need to solve circular dependency
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
+)
+
+type piboot struct {
+	rootdir string
+	basedir string
+}
+
+func (p *piboot) setDefaults() {
+	p.basedir = "/boot/piboot/"
+}
+
+func (p *piboot) processBlOpts(blOpts *Options) {
+	if blOpts == nil {
+		return
+	}
+
+	switch {
+	case blOpts.Role == RoleRecovery || blOpts.NoSlashBoot:
+		if !blOpts.PrepareImageTime {
+			p.rootdir = ubuntuSeedDir
+		}
+		// RoleRecovery or NoSlashBoot imply we use
+		// the environment file in /piboot/ubuntu as
+		// it exists on the partition directly
+		p.basedir = pibootPartFolder
+	}
+}
+
+// newPiboot creates a new Piboot bootloader object
+func newPiboot(rootdir string, blOpts *Options) Bootloader {
+	p := &piboot{
+		rootdir: rootdir,
+	}
+	p.setDefaults()
+	p.processBlOpts(blOpts)
+	return p
+}
+
+func (p *piboot) Name() string {
+	return "piboot"
+}
+
+func (p *piboot) dir() string {
+	if p.rootdir == "" {
+		panic("internal error: unset rootdir")
+	}
+	return filepath.Join(p.rootdir, p.basedir)
+}
+
+func (p *piboot) envFile() string {
+	return filepath.Join(p.dir(), pibootCfgFilename)
+}
+
+// piboot enabled if env file exists
+func (p *piboot) Present() (bool, error) {
+	return osutil.FileExists(p.envFile()), nil
+}
+
+// Variables stored in ubuntu-seed:
+//   snapd_recovery_system
+//   snapd_recovery_mode
+//   snapd_recovery_kernel
+// Variables stored in ubuntu-boot:
+//   kernel_status
+//   snap_kernel
+//   snap_try_kernel
+//   snapd_extra_cmdline_args
+//   snapd_full_cmdline_args
+//   recovery_system_status
+//   try_recovery_system
+func (p *piboot) SetBootVars(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	dirty := false
+	needsReconfig := false
+	pibootConfigureAllow := true
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		// "kernel_status" transitions from "try" to "trying"/"" are
+		// coming from the initramfs (simulating what the bootloader
+		// should do) and do not retrigger a configuration.
+		if k == "kernel_status" && env.Get(k) == "try" {
```

Comment:
I don't understand the comment completely, why is the old value relevant instead of the new one? this seems too much of an implicit logic to me. Either we need a mode or this needs to be based on the value of v, not the old value

Reply:
The old value is relevant as we want to catch transitions from "try" to "trying" or "". These transitions come only from the initramfs (which is simulating for piboot what grub.cfg would do), and in that case we want to avoid re-generating yet the piboot configuration files - we want that only to be done after snapd rootfs has started and it has considered that the new kernel snap installation was successful.

I think that an alternative for this would be maybe to add some method (`SetBootVarsFromInitramfs()`?) in an interface only for piboot, that would allow getting rid of this.